source/slang/slang-ir-variable-scope-correction.cpp

```diff
@@ -0,0 +1,317 @@
+            // 2. Declare the dstPtr at beginning of the function, so it's globally available
+            // 3. Replace the operand with the dstPtr
+            if (auto type = operand->getDataType())
+            {
+                m_builder.setInsertBefore(insertLoc);
+                auto dstPtr = m_builder.emitVar(type);
+
+                m_builder.setInsertAfter(operand);
+                m_builder.emitStore(dstPtr, operand);
+
+                clonedInst->setOperand(i, dstPtr);
+            }
+        }
+    }
+
+    // last, because the user is still referencing the original instruction, we need to
+    // replace operands in the use site instruction with the cloned instruction
+    _replaceOperand(user, inst, clonedInst);
+}
+
+// For storable instructions, we don't want to duplicate it at use site, so we just need to store the result after the instruction
+void VariableScopeCorrectionContext::_processStorableInst(IRInst* insertLoc, IRInst* inst, IRInst* user)
+{
+    auto type = inst->getDataType();
+    // store instruction must have a result type
+    SLANG_ASSERT(type);
+
+    // declare a new variable at the beginning of the function used to store the result of the instruction
+    m_builder.setInsertBefore(insertLoc);
+    auto dstPtr = m_builder.emitVar(type);
+
+    // insert a store instruction after the instruction
+    m_builder.setInsertAfter(inst);
+    m_builder.emitStore(dstPtr, inst);
+
+    // last, replace operands in the use site instruction with the new variable
+    // Note, because "dstPtr" is a pointer type, we have to insert a load(dstPtr) instruction before use it.
+    // Simply replace any operand with pointer could generate error code.
+    m_builder.setInsertBefore(user);
+    auto loadInst = m_builder.emitLoad(dstPtr->getDataType(), dstPtr);
+    _replaceOperand(user, inst,  loadInst);
+}
+
+void  VariableScopeCorrectionContext::_replaceOperand(IRInst* inst, IRInst* oldOperand, IRInst* newOperand)
+{
+    // TODO: I'd like to use "user->replaceUsesWith(clonedInst);" here, but it seems not replace the operands at all.
+    SlangUInt operandCount = inst->getOperandCount();
+
+    // Traverse all the operands to find out which one is the invisible instruction, and replace it with the new one.
+    for (SlangUInt i = 0; i < operandCount; i++)
+    {
+        auto operand = inst->getOperand(i);
+
+        if (operand == oldOperand)
+        {
+            inst->setOperand(i, newOperand);
+        }
+    }
+}
+
+bool VariableScopeCorrectionContext::_isStorableInst(IRInst* inst)
+{
+    auto type = inst->getDataType();
+
+    // If the instruction has a result, and the result is not a void type, we consider it as a storable instruction
+    if (type)
+    {
+        // Take care of pointer type, because we can't store a pointer type, so pointer type is regarded as a non-storable instruction
+        if (type->getOp() == kIROp_PtrType)
```

Comment:
Instead of checking what can't be stored, just check for what can be stored.
There are way more things we can't store than the things we can.

We can only store these:
IRBasicType
IRVectorType
IRMatrixType
IRArrayTypeBase, if element type is storable.
IRStructType

You can define a `isStorableType(IRType* type)` function to compute this.

Reply:
Also, everything is storable if we are on cpp/cuda target.

Reply:
Yea, I did the update in my latest change.
I will add the cpp/cuda check then. But the good news is that after I fix the last issue in the `_processStorableInst` that fails the Falcor image test, now all the CIs passed.

Reply:
The cpp/cuda check is done.